pyi.py

```diff
@@ -162,6 +162,20 @@ class TypeVarInfo(NamedTuple):
 
 
 class PyiAwareFlakesChecker(FlakesChecker):
```

Comment:
Here's a patch that fixes the issue with recursive definitions:

```suggestion
class PyflakesPreProcessor(ast.NodeTransformer):
    """Transform AST prior to passing it to pyflakes.

    This reduces false positives on recursive class definitions.
    """

    def visit_ClassDef(self, node: ast.ClassDef) -> ast.ClassDef:
        self.generic_visit(node)
        new_node = deepcopy(node)
        new_node.bases = [
            base.value if isinstance(base, ast.Subscript) else base
            for base in new_node.bases
        ]
        return new_node


class PyiAwareFlakesChecker(FlakesChecker):
    def __init__(self, tree: ast.AST, filename: str) -> None:
        return super().__init__(PyflakesPreProcessor().visit(tree), filename)

```

Reply:
Definitely seems like the easiest option here, though we are giving up `F821` (and other pyflakes checks) for anything inside the bases, for example:
```python
class Foo(list[Bar, Undefined]): ...
```
if `Undefined` does not exist in the file (forgotten import or a typo), it won't be reported (I know that in typeshed `F821` is currently disabled but still probably not the ideal solution)

Reply:
Right now, F821 has an unacceptable level of false positives for stub files, even with the monkeypatching that flake8-pyi does to pyflakes. So I see the first priority as getting the level of F821 false positives down to a level where it's usable in stub files. If we introduce a few false negatives in the process of reducing the number of false negatives, I think that's fine for now, since the vast majority of users of stub files will probably (hopefully!) be running a type checker on their stub files in CI.

Reply:
So with the `annotationsFutureEnabled` enabled in pyflakes I see only `17` F821 errors in `typeshed/stdlib` (compared to 87 on master), most of them having to do with the wrong class def order (Basically the same errors from here https://github.com/PyCQA/flake8-pyi/pull/364#issuecomment-1507202331). If I apply this on master, w/o the class def fixes, I get 0 `F821` errors in `stdlib`. I still think your suggestion is probably the best solution and I couldn't think of anything better that doesn't involve full-blown cycle detection which gets complex really quick. Though if `F821` gets reenabled, we'll have to come up with something else eventually. What do you think? Might be worth opening a separate PR with the annotations fixes since it is kind of unrelated to the class defs anyway..

Reply:
> If I apply this on master, w/o the class def fixes, I get 0 F821 errors in stdlib.

If I do the following:
1. Check out your PR locally
2. Apply the patch above to my local copy of your branch
3. Apply this diff to my local clone of typeshed:

    ```diff
    --- a/.flake8
    +++ b/.flake8
    @@ -36,14 +36,14 @@
     extend-ignore = A, D, N8, SIM, RST, TYP, E301, E302, E305, E501
     per-file-ignores =
       *.py: E203
    -  *.pyi: B, E701, E741, F401, F403, F405, F821, F822
    +  *.pyi: B, E701, E741, F401, F403, F405, F822
       # Since typing.pyi defines "overload" this is not recognized by flake8 as typing.overload.
       # Unfortunately, flake8 does not allow to "noqa" just a specific error inside the file itself.
       # https://github.com/PyCQA/flake8/issues/1079
       #     F811 redefinition of unused '...'
    -  stdlib/typing.pyi: B, E701, E741, F401, F403, F405, F811, F821, F822
    +  stdlib/typing.pyi: B, E701, E741, F401, F403, F405, F811, F822
       # Generated protobuf files include docstrings
    -  *_pb2.pyi: B, E701, E741, F401, F403, F405, F821, F822, Y021, Y026, Y053, Y054
    +  *_pb2.pyi: B, E701, E741, F401, F403, F405, F822, Y021, Y026, Y053, Y054
    ```

4. Run flake8 on `typeshed/stdlib`

Then I get this output:

```
stdlib\enum.pyi:212:12: F821 undefined name 'IntFlag'
stdlib\enum.pyi:240:49: F821 undefined name 'KEEP'
stdlib\multiprocessing\synchronize.pyi:17:24: F821 undefined name 'Semaphore'
stdlib\multiprocessing\synchronize.pyi:39:12: F821 undefined name 'SemLock'
stdlib\multiprocessing\synchronize.pyi:42:13: F821 undefined name 'SemLock'
stdlib\multiprocessing\synchronize.pyi:45:17: F821 undefined name 'SemLock'
stdlib\sqlite3\dbapi2.pyi:393:17: F821 undefined name 'DatabaseError'
stdlib\sqlite3\dbapi2.pyi:394:21: F821 undefined name 'Error'
```

If I then run flake8 on `typeshed/stubs`, I get this output:

```
stubs\openpyxl\openpyxl\utils\datetime.pyi:6:40: NQA102 "# noqa: F821" has no matching violations
stubs\openpyxl\openpyxl\utils\datetime.pyi:7:32: NQA102 "# noqa: F821" has no matching violations
stubs\passlib\passlib\apps.pyi:23:36: NQA102 "# noqa: F821" has no matching violations
stubs\passlib\passlib\apps.pyi:29:33: NQA102 "# noqa: F821" has no matching violations
stubs\passlib\passlib\apps.pyi:35:38: NQA102 "# noqa: F821" has no matching violations
stubs\protobuf\google\protobuf\compiler\plugin_pb2.pyi:128:155: NQA102 "# noqa: F821" has no matching violations
stubs\protobuf\google\protobuf\descriptor_pb2.pyi:255:148: NQA102 "# noqa: F821" has no matching violations
stubs\protobuf\google\protobuf\descriptor_pb2.pyi:336:150: NQA102 "# noqa: F821" has no matching violations
stubs\protobuf\google\protobuf\descriptor_pb2.pyi:657:155: NQA102 "# noqa: F821" has no matching violations
stubs\protobuf\google\protobuf\descriptor_pb2.pyi:919:142: NQA102 "# noqa: F821" has no matching violations
stubs\protobuf\google\protobuf\descriptor_pb2.pyi:936:144: NQA102 "# noqa: F821" has no matching violations
stubs\protobuf\google\protobuf\descriptor_pb2.pyi:1170:165: NQA102 "# noqa: F821" has no matching violations
stubs\protobuf\google\protobuf\type_pb2.pyi:125:133: NQA102 "# noqa: F821" has no matching violations
stubs\protobuf\google\protobuf\type_pb2.pyi:212:147: NQA102 "# noqa: F821" has no matching violations
stubs\setuptools\pkg_resources\__init__.pyi:76:28: F821 undefined name 'Distribution'
stubs\setuptools\pkg_resources\__init__.pyi:136:20: F821 undefined name 'NullProvider'
stubs\setuptools\pkg_resources\__init__.pyi:136:34: F821 undefined name 'IResourceProvider'
stubs\setuptools\pkg_resources\__init__.pyi:136:53: F821 undefined name 'IMetadataProvider'
stubs\tensorflow\tensorflow\compiler\xla\service\hlo_pb2.pyi:806:156: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\compiler\xla\service\hlo_pb2.pyi:1079:156: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\compiler\xla\xla_data_pb2.pyi:851:155: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\compiler\xla\xla_data_pb2.pyi:1310:160: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\compiler\xla\xla_data_pb2.pyi:1418:138: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\compiler\xla\xla_data_pb2.pyi:1576:153: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\framework\api_def_pb2.pyi:50:146: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\framework\dataset_options_pb2.pyi:135:162: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\framework\graph_transfer_info_pb2.pyi:196:159: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\config_pb2.pyi:348:146: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\config_pb2.pyi:375:164: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\config_pb2.pyi:686:182: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\config_pb2.pyi:1086:150: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\core_platform_payloads_pb2.pyi:32:158: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\data_service_pb2.pyi:59:165: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\data_service_pb2.pyi:163:161: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\rewriter_config_pb2.pyi:70:146: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\rewriter_config_pb2.pyi:118:152: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\rewriter_config_pb2.pyi:135:168: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\rewriter_config_pb2.pyi:154:154: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\saved_object_graph_pb2.pyi:475:152: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\saver_pb2.pyi:29:174: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\struct_pb2.pyi:334:159: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\tensor_bundle_pb2.pyi:44:157: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\tpu\compilation_result_pb2.pyi:36:160: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\tpu\optimization_parameters_pb2.pyi:738:158: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\tpu\optimization_parameters_pb2.pyi:805:159: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\tpu\optimization_parameters_pb2.pyi:846:161: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\tpu\topology_pb2.pyi:31:170: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\tpu\tpu_embedding_configuration_pb2.pyi:30:153: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\tpu\tpu_embedding_configuration_pb2.pyi:51:177: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\protobuf\verifier_config_pb2.pyi:29:146: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\util\event_pb2.pyi:151:140: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\util\event_pb2.pyi:202:156: NQA102 "# noqa: F821" has no matching violations
stubs\tensorflow\tensorflow\core\util\test_log_pb2.pyi:486:157: NQA102 "# noqa: F821" has no matching violations
```

That looks pretty promising to me! It picks up a bunch of true positives where the classes are being defined in the wrong order, but doesn't have any false positives that I'm aware of. In fact, it fixes many false positives that flake8 currently emits on the typeshed codebase, which we're currently manually `# noqa`'ing. It improves the experience for users of flake8-pyi, without introducing any regressions.

I would just go with this patch :)